File: Kotlin/src/main/kotlin/Day3.kt
package day3

import utils.*
typealias TreeMap = List<String>
typealias MapCell = Char
const val tree: MapCell = '#'

typealias Point = Pair<Int, Int>
val f1 = { xy: Point -> Point(xy.first + 1, xy.second + 1 )}
val f2 = { xy: Point -> Point(xy.first + 3, xy.second + 1 )}
val f3 = { xy: Point -> Point(xy.first + 5, xy.second + 1)}
val f4 = { xy: Point -> Point(xy.first + 7, xy.second + 1 )}
val f5 = { xy: Point -> Point(xy.first + 1, xy.second + 2)}
val allFs = listOf(f1, f2, f3, f4, f5)

fun itemAt( xy: Point, tm: TreeMap): Char {
    if (xy.second > tm.size) {
        return '.'
    }
    return  tm[xy.second ][xy.first % 31]
}

fun allPoints(size: Int, stepF: (Point) -> Point): List<Point> =
     (0 until size).fold(mutableListOf(Point(0,0))){ acc, _ ->
         acc.add(stepF(acc.last()))
         acc}

fun part1(treeMap: List<String>, stepFunc: (Point) -> Point): Int{
    val all = allPoints(treeMap.size - 1) { stepFunc(it) }
    return all.map { itemAt(it, treeMap)}
            .filter { it == tree }
            .count()
}
fun part2(treeMap: List<String>): Long{

    return allFs.map {part1(treeMap, it)}
            .fold(1, {acc, x -> x * acc })
}

fun main(args: Array<String>) {
    val treeMap = readFileAsLines("resources/Day3Part1.txt")
    println(part1(treeMap, f2))
    print(part2(treeMap))
}

File: Haskell/AllDays/src/Day3.hs
module Day3 where

import Control.Monad
import Data.List

type Point = (Int, Int)
type TreeMap = [String]
tree = '#'
f1 (x, y) = (x + 1, y + 1)
f2 (x, y) = (x + 3, y + 1)
f3 (x, y) = (x + 5, y + 1)
f4 (x, y) = (x + 7, y + 1)
f5 (x, y) = (x + 1, y + 2)
funcs = [f1, f2, f3, f4, f5]

itemAt :: Point -> TreeMap -> Char
itemAt (x, y) tm  
    | y >= length tm = '.' 
    | otherwise = (tm !! y) !! (x `mod` 31)

allPoints :: Int -> (Point -> Point) -> [Point]  
allPoints n f = foldr(\x acc -> f (head acc) : acc  ) [(0,0)] [1..n]


part1 :: TreeMap -> (Point -> Point) -> Int
part1 tm f = res where
    size = length tm - 1
    res = length . filter (== tree) . map (`itemAt` tm) $ allPoints size f

part2 :: TreeMap -> Int
part2 tm = product . map (part1 tm) $ funcs 
 
day3Main :: IO () 
day3Main = do
    content <- readFile "data/Day3Part1.txt"
    print $ part1 (lines content) f2
    print $ part2 (lines content)
   